# Chances of winning

- What are the chances of winning main price in 6/49 lottery?
- Is it worth to buy more tickets to have higher chance of winning?
- Maybe there are much bigger chances of getting 5/49?

In this project we are going to answer all that questions. 
Main goal for today is to show gambling addicts what are the odds to win any money. Let's get started.

First of all, we will go with writing two functions: <br>
 - `factorial()` — a function that calculates factorials
    

![factorial](C:/Users/gener/Desktop/factorial.png)


- `combinations()` — a function that calculates combinations

![combinations](C:/Users/gener/Desktop/combinations.png)

In [1]:
import pandas as pd
import numpy as np

import statistics

In [2]:
def factorial(n):
    result = 1
    for i in range(n, 0, -1):#loop goes from n to 0(not included) decreasing by 1
        result *= i
    return result
def combinations(n,k):
    diff = n-k
    result = factorial(n)/(factorial(k)*factorial(diff))
    #we are using factorial function here-according to the formula
    return result    

In [3]:
print(factorial(5))

120


In [4]:
print(combinations(49,6))

13983816.0


Everything working just fine for now. Let's keep going!

In [5]:
def one_ticker_probability(user_numbers):
    total_outcomes = combinations(49,6)
    probability = 1/total_outcomes
    percentage_probability = probability * 100
    return print('''Your chances to win the big prize with the numbers {}\nare {:.7f}%.
In other words, you have a 1 in {:,} chances to win.'''.format(user_numbers,
                    percentage_probability, int(total_outcomes)))

We now test a function with two different outputs:

In [6]:
print(one_ticker_probability([1,2,3,4,5,6]))

Your chances to win the big prize with the numbers [1, 2, 3, 4, 5, 6]
are 0.0000072%.
In other words, you have a 1 in 13,983,816 chances to win.
None


In [7]:
print(one_ticker_probability([8, 13, 21, 39,49]))

Your chances to win the big prize with the numbers [8, 13, 21, 39, 49]
are 0.0000072%.
In other words, you have a 1 in 13,983,816 chances to win.
None


In [8]:
canada_winnings = pd.read_csv('649.csv')
canada_winnings.shape

(3665, 11)

In [9]:
canada_winnings.head(3)

PRODUCT  DRAW NUMBER  SEQUENCE NUMBER  DRAW DATE  NUMBER DRAWN 1  \
0      649            1                0  6/12/1982               3   
1      649            2                0  6/19/1982               8   
2      649            3                0  6/26/1982               1   

   NUMBER DRAWN 2  NUMBER DRAWN 3  NUMBER DRAWN 4  NUMBER DRAWN 5  \
0              11              12              14              41   
1              33              36              37              39   
2               6              23              24              27   

   NUMBER DRAWN 6  BONUS NUMBER  
0              43            13  
1              41             9  
2              39            34

In [10]:
canada_winnings.tail(3)

PRODUCT  DRAW NUMBER  SEQUENCE NUMBER  DRAW DATE  NUMBER DRAWN 1  \
3662      649         3589                0  6/13/2018               6   
3663      649         3590                0  6/16/2018               2   
3664      649         3591                0  6/20/2018              14   

      NUMBER DRAWN 2  NUMBER DRAWN 3  NUMBER DRAWN 4  NUMBER DRAWN 5  \
3662              22              24              31              32   
3663              15              21              31              38   
3664              24              31              35              37   

      NUMBER DRAWN 6  BONUS NUMBER  
3662              34            16  
3663              49             8  
3664              48            17

Above we have first and last three rows from Canada winning draws between 1982-2018!

In [11]:
def extract_numbers(row):
    row = row[4:10]
    row = set(row.values)
    return row

winning_numbers = canada_winnings.apply(extract_numbers, axis=1)
winning_numbers.head()

0    {3, 41, 11, 12, 43, 14}
1    {33, 36, 37, 39, 8, 41}
2     {1, 6, 39, 23, 24, 27}
3     {3, 9, 10, 43, 13, 20}
4    {34, 5, 14, 47, 21, 31}
dtype: object

In [12]:
def check_historical_occurrence(user_numbers, historical_numbers):   
    '''
    user_numbers: a Python list
    historical numbers: a pandas Series
    '''
    
    user_numbers_set = set(user_numbers)
    check_occurrence = historical_numbers == user_numbers_set
    n_occurrences = check_occurrence.sum()
    
    if n_occurrences == 0:
        print('''The combination {} has never occured.
This doesn't mean it's more likely to occur now. Your chances to win the big prize in the next drawing using the combination {} are 0.0000072%.
In other words, you have a 1 in 13,983,816 chances to win.'''.format(user_numbers, user_numbers))
        
    else:
        print('''The number of times combination {} has occured in the past is {}.
Your chances to win the big prize in the next drawing using the combination {} are 0.0000072%.
In other words, you have a 1 in 13,983,816 chances to win.'''.format(user_numbers, n_occurrences,
                                                                            user_numbers))

In [13]:
test_input_3 = [33, 36, 37, 39, 8, 41]
check_historical_occurrence(test_input_3, winning_numbers)

The number of times combination [33, 36, 37, 39, 8, 41] has occured in the past is 1.
Your chances to win the big prize in the next drawing using the combination [33, 36, 37, 39, 8, 41] are 0.0000072%.
In other words, you have a 1 in 13,983,816 chances to win.


In [14]:
test_input_4 = [2, 9, 16, 24, 40, 49]
check_historical_occurrence(test_input_4, winning_numbers)

The combination [2, 9, 16, 24, 40, 49] has never occured.
This doesn't mean it's more likely to occur now. Your chances to win the big prize in the next drawing using the combination [2, 9, 16, 24, 40, 49] are 0.0000072%.
In other words, you have a 1 in 13,983,816 chances to win.


We have tested two different samples and obviously chances of winnings are the same no matter which number player picks.
 - but how buying more tickets influence on chances of winning?

In [66]:
def multi_ticket_probability(n):
    total_outcomes = combinations(49,6)
    probability = n/total_outcomes
    percentage = probability * 100
    if n == 0:
        print('There is no way you to win')
    if n ==1:
        print("Your chances of winning with 1 ticket are: {:.7f}%".format(percentage))
    if n>1:
        print("Your chances of winning with {} tickets are: {:.7f}%".format(n, percentage))
    
samples = [1, 10, 100, 10000, 1000000, 6991908, 13983816]
for sample in samples:
    multi_ticket_probability(sample)
    

Your chances of winning with 1 ticket are: 0.0000072%
Your chances of winning with 10 tickets are: 0.0000715%
Your chances of winning with 100 tickets are: 0.0007151%
Your chances of winning with 10000 tickets are: 0.0715112%
Your chances of winning with 1000000 tickets are: 7.1511238%
Your chances of winning with 6991908 tickets are: 50.0000000%
Your chances of winning with 13983816 tickets are: 100.0000000%


The last function will give us % of winning with 2, 3, 4 or 5 numbers.

In [38]:
def probability_less_6(n):
    n_combinations_tickets = combinations(6, n)
    n_outcomes = combinations(43, 6-n)
    successful_outcomes = n_combinations_tickets * n_outcomes
    
    total_outcomes = combinations(49,6)
    probability = successful_outcomes / total_outcomes
    
    percentages = round(probability * 100, 6)
    return ('Chances of winning with {} out of 6 numbers are {}%'.format(n, percentages))


In most lotteries, even if the player did not guess 6/6 numbers, he still can win some money by guessing 2/6, 3/6, 4/6 or 5/6. Sounds good enough huh? But percentages will be slighly or much better in favor of the player?<br>
Let's check this out.

In [39]:
possible_inputs = [2,3,4,5]

for tickets in possible_inputs:
    chances = probability_less_6(tickets)
    print(chances)
    print('---------------------------------------------------------------------')

Chances of winning with 2 out of 6 numbers are 13.237803%
---------------------------------------------------------------------
Chances of winning with 3 out of 6 numbers are 1.76504%
---------------------------------------------------------------------
Chances of winning with 4 out of 6 numbers are 0.096862%
---------------------------------------------------------------------
Chances of winning with 5 out of 6 numbers are 0.001845%
---------------------------------------------------------------------


So now, thanks to 'probability_less_6' function, user can see what percentages are to win with 'n' out of 6 numbers. 
As you can notice, winning with 3 out of 6 numbers are over 1%!. But for example in Poland you would win for that 24PLN(~6$). 

# Summarize

Let's answer our questions first.<br>
- The chances of winning main price are: 0.0000072%
- To have at least 1% to win, player has to buy 139,839 tickets. Is it worh it? Numbers speak for themselves
- Chances of winning with 5/49 are 256 times bigger! But it's still: 0.001845% to win. 

<center>In conclusion, it is each individual's business whether it is worth playing or not. I just showed what the chances are for it, and leave it to you to think about it.<br>